velox/docs/functions/presto/aggregate.rst

```diff
@@ -143,6 +143,17 @@ General Aggregate Functions
     The final state is returned. Throws an error if ``initialState`` is NULL or
     ``inputFunction`` or ``combineFunction`` returns a NULL.
 
+    Take care when designing ``initialState``, ``inputFunction`` and ``combineFunction``.
+    These need to support evaluating aggregation in a distributed manner using partial
+    aggregation on many nodes, followed by shuffle over group-by keys, followed by
+    final aggregation. Make sure that
+
+     combineFunction(s1, s2) = combineFunction(s2, s1) for any s1 and s2;
```

Comment:
These properties are called Idempotent (https://en.wikipedia.org/wiki/Idempotence) and associative (https://en.wikipedia.org/wiki/Associative_property) in most docs. We could use those terms for clarity.

Reply:
I'm not sure most reader would find it easy to understand. BTW, I don't think these terms apply here.

> combineFunction(s1, s2) = combineFunction(s2, s1)

This is not idempotence, rather symmetry. In some contexts, initialState is called the "identity" element.

> combineFunction(initialState, initialState) = initialState

I think this is a requirement as well. This one looks like idempotence, but applies only to one value, not all values.

Associativity applies to one operation, but here we have 2: inputFunction + combineFunction. The requirements above is about interaction between these 2 functions.

That said, combineFunction(s, combineFunction(u, v)) = combineFunction(combineFunction(s, u), v) should be true. This is associativity of combineFunction.

Reply:
@mbasmanova : Agree that because of having inputFunction + combineFunction makes the writing more complex.

The idempotence was for initial function applied on initial state twice should return the same result.

You are right that symmetry is required as well The symmetry is called commutative https://en.wikipedia.org/wiki/Commutative_property

But in general I'm fine with not using those terminologies.